statistics/handle/handle.go

```diff
@@ -119,6 +119,284 @@ type Handle struct {
 	sysProcTracker sessionctx.SysProcTracker
 	// serverIDGetter is used to get server ID for generating auto analyze ID.
 	serverIDGetter func() uint64
+	// tableLocked used to store locked tables
+	tableLocked []int64
+}
+
+// GetTableLockedAndClear for unit test only
+func (h *Handle) GetTableLockedAndClear() []int64 {
+	tableLocked := h.tableLocked
+	h.tableLocked = make([]int64, 0)
+	return tableLocked
+}
+
+// LoadLockedTables load locked tables from store
+func (h *Handle) LoadLockedTables() error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
+	rows, _, err := h.execRestrictedSQL(ctx, "select table_id from mysql.stats_table_locked")
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	h.tableLocked = make([]int64, len(rows))
+	for i, row := range rows {
+		h.tableLocked[i] = row.GetInt64(0)
+	}
+
+	// update SessionStatsCollector
+	{
+		prev := h.listHead
+		prev.Lock()
+		for curr := prev.next; curr != nil; curr = curr.next {
+			curr.Lock()
+			curr.tableLocked = h.tableLocked
+			curr.Unlock()
+		}
+		prev.Unlock()
+	}
+	return nil
+}
+
+// AddLockedTables add locked tables id  to store
+func (h *Handle) AddLockedTables(tids []int64, tables []*ast.TableName) (string, error) {
+	var err error
+
+	h.mu.Lock()
+	defer h.mu.Unlock()
+
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnStats)
+
+	exec := h.mu.ctx.(sqlexec.SQLExecutor)
+
+	if len(tids) > 1 {
+		_, err = exec.ExecuteInternal(ctx, "begin pessimistic")
+		if err != nil {
+			return "", err
+		}
+	}
+
+	//load tables to check duplicate when insert
+	rows, _, err := h.execRestrictedSQL(ctx, "select table_id from mysql.stats_table_locked")
+	if err != nil {
+		return "", err
+	}
+
+	dupTables := make([]string, 0)
+	tableLocked := make([]int64, 0)
+	for _, row := range rows {
+		tableLocked = append(tableLocked, row.GetInt64(0))
+	}
+
+	strTids := fmt.Sprintf("%v", tids)
+	logutil.BgLogger().Info("[stats] lock table ", zap.String("tableIDs", strTids))
+	for i, tid := range tids {
+		_, err = exec.ExecuteInternal(ctx, "insert into mysql.stats_table_locked(table_id) select %? from dual where not exists(select table_id from mysql.stats_table_locked where table_id = %?)", tid, tid)
+		if err != nil {
+			logutil.BgLogger().Error("[stats] error occurred when insert mysql.stats_table_locked ", zap.Error(err))
+			return "", err
+		}
+		// update handle
+		if !isTableLocked(tableLocked, tid) {
+			tableLocked = append(tableLocked, tid)
+		} else {
+			dupTables = append(dupTables, tables[i].Schema.L+"."+tables[i].Name.L)
+		}
+	}
+
+	if len(tids) > 1 {
+		err = finishTransaction(ctx, exec, err)
```

Comment:
Maybe it is better to `defer finishTransaction` right after `begin begin pessimistic`.

Reply:
if finishTransaction get failed, 'mysql.stats_table_locked' won't be updated, and "h.tableLocked = tableLocked" won't be execute, so h.tableLocked is consistent with table.
if change to defer finishTransaction , if finishTransaction get failed, table won't be updated but h.tableLocked will be updated,  data in table and memory is inconsistent.